src/lib.rs

```diff
@@ -40,7 +52,25 @@ pub unsafe extern "C" fn pthread_create(
     // We make this closure in the first place because threadCreate expects a void return type, but
     // entrypoint returns a pointer so the types are incompatible.
     let main: *mut Box<dyn FnOnce() -> *mut libc::c_void> =
-        Box::into_raw(Box::new(Box::new(move || entrypoint(value))));
+        Box::into_raw(Box::new(Box::new(move || {
+            let ret = entrypoint(value);
+
+            // destruct the THREAD_SELF variable, if it was ever created
+            if let Some(p) = THREAD_SELF {
+                let _ = Box::from_raw(p);
+            }
+
+            // Destruct thread local values where possible. We can't guarantee destructors
+            // will run for all values, which is similar to other Unix-like platforms:
+            // https://doc.rust-lang.org/std/thread/struct.LocalKey.html#platform-specific-behavior
+            for (&key, &value) in &LOCALS {
```

Comment:
Good point. Actually, I noticed that the segfault when using `#[thread_local]` (https://github.com/Meziu/rust-horizon/pull/16) doesn't occur without this change, since the dtors aren't run. So probably worth hashing out separately... I will back it out of this PR.